front/src/modules/simulationResult/components/SpaceTimeChart/createSpaceTimeChart.ts

```diff
@@ -2,64 +2,68 @@ import * as d3 from 'd3';
 import { select as d3select } from 'd3-selection';
 
 import { Chart, SimulationTrain, ConsolidatedPosition } from 'reducers/osrdsimulation/types';
-import { ChartAxes } from 'modules/simulationResult/consts';
+import { CHART_AXES } from 'modules/simulationResult/consts';
 import {
   defineLinear,
   defineTime,
-  isSpaceTimeChart,
 } from 'modules/simulationResult/components/ChartHelpers/ChartHelpers';
 import defineChart from 'modules/simulationResult/components/ChartHelpers/defineChart';
+import { SPACE_TIME_CHART_ID } from './consts';
 
-export default function createSpaceTimeChart<T extends number | Date>(
+function getMaxPosition(trains: SimulationTrain[], pos: 'tailPosition' | 'headPosition') {
+  return d3.max(
+    trains.flatMap(
+      (train) =>
+        d3.max(
+          train[pos].map(
+            (section) =>
+              d3.max(section.map((step: ConsolidatedPosition) => step.position)) as number
+          )
+        ) as number
+    )
+  ) as number;
+}
+
+export default function createSpaceTimeChart(
   chart: Chart | undefined,
-  chartID: string,
-  dataSimulation: SimulationTrain[],
-  heightOfSpaceTimeChart: number,
-  keyValues: ChartAxes,
+  trains: SimulationTrain[],
+  chartHeight: number,
   ref: React.MutableRefObject<HTMLDivElement> | React.RefObject<HTMLDivElement>,
   reset: boolean,
   rotate: boolean
 ): Chart {
-  d3select(`#${chartID}`).remove();
+  d3select(`#${SPACE_TIME_CHART_ID}`).remove();
 
-  const xValues: T[] = dataSimulation
+  const resetChart = chart === undefined || reset;
+
+  const timeValues = trains
     .map((train) =>
```

Comment:
Can't we use .flatMap() as we have a .map() and a .flat() ?

```suggestion
   .flatMap((train) =>
    (train.eco_headPosition ?? train.headPosition).flatMap((section) =>
      section.map((position) => position.time)
    )
  ) as Date[];
```

Reply:
flatMap is more direct, you can read documentation about it here : https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flatMap :blush:
As explained in the doc, flatMap() is slightly more efficient than map().flat()

Reply:
Thanks, I misunderstood ! Done :white_check_mark: